Transactor Preclaim validation

ximinez opened the pull request:
Sort of as a continuation of the `preflight` checks we implemented in #1198, this implements `preclaim` checking.

* Unlike `precheck`, which does all the static checks that do not need a view, `preclaim` will do all of the static checks that _do_ require a view.
* If `preclaim` succeeds, then the transaction is very likely to claim a fee. This will determine in the transaction is safe to relay without being applied to the open ledger (ie. it can be put into the forthcoming transaction queue).
  * "Succeeds" in this case is defined as returning a `tes` or `tec`, since both lead to claiming a fee.
* It will also return base fee information, which will be useful for forthcoming fee escalation functionality.
* Finally, the interfaces for the `preflight`, `preclaim`, and `doApply` have been modified so that the output of one must be passed as input to the next. This is not strictly necessary, but it actually simplifies the calling a little bit, and makes it less likely that a necessary step gets bypassed.
* For existing callers, the `ripple::apply` interface has not changed one bit, though the implementation obviously has.

I intentionally left the changes as several [FOLD] commits, most of which change one small piece of the puzzle in the hope of making the review easier. If that makes it harder instead, let me know and I'll squash.

Reviewers: @vinniefalco, @scottschurr 

I'll be out from 9/23 - 9/27, but will try to keep an eye on email and Slack mentions in case there are questions.

ximinez commented:
Rebased against `0.29.1-rc1` and resolved conflicts.

scottschurr commented:
Looks good.  I have a question I should have asked at the start, but I'm really confident you have a good answer.  So, we've got all of this checking distributed into pre-flight, pre-claim, and apply.  The pre-flight and pre-claim checks are good for knowing whether a transaction should be distributed to the network.  And transactions that pass pre-flight and pre-claim may (eventually) be put in a queue for later handling (based on your Fee work).

So here's the question.  If a transaction passes pre-claim and is put in a queue, then eventually we pull the transaction out of the queue, When the transaction leaves the queue, then we have to do the pre-flight and pre-claim checks over again, right?  After all, we expect that the state of the ledger has changed.  I'm just checking that there isn't any kind of work reduction going on here that I missed.

scottschurr commented:
:+1: I'm pretty sure that any small concerns I raised have been addressed.  Looks good.

ximinez commented:
>  If a transaction passes pre-claim and is put in a queue, then eventually we pull the transaction out of the queue, When the transaction leaves the queue, then we have to do the pre-flight and pre-claim checks over again, right?

Optimally, we'll only need to do the `preflight` (including the signature check) if the `Rules` have changed, something that's happens very infrequently. But we will have to repeat the `preclaim` regardless. Being able to skip `preflight` is a pretty big win, though, since that's where we check the signature.

scottschurr commented:
Nice.  Thanks for the explanation.

ximinez commented:
You're welcome.

vinniefalco commented on src/ripple/app/tx/apply.h:
```diff
@@ -39,11 +41,24 @@ class Application;
     @return The TER code (a `tem` or tesSUCCESS)
 */
-TER
+PreflightResult
```
This doesn't make sense - a public API function returning a private struct? (`PreflightResult` is declared in `impl/Transactor.h`, which is a private header).

ximinez replied:
Good catch. That was not my intention.
Option 1: Make the `PreflightResult` and `PreclaimResult` structs public, and make the context member of the struct private.
Option 2: Take `preflight, preclaim, and doApply` out of the public API, and put them under `impl` as well.

I'm leaning toward Option 2 because I've got the transaction queue under `tx/impl` right now, so there is no need for those functions to be public.

vinniefalco replied:
2 is better if they aren't called externally

ximinez replied:
Done.

vinniefalco commented on src/ripple/app/tx/impl/CreateOffer.cpp:
```diff
@@ -569,22 +645,18 @@ CreateOffer::applyGuts (ApplyView& view, ApplyView& view_cancel)
 
-    STAmount saTakerPays = tx().getFieldAmount (sfTakerPays);
-    STAmount saTakerGets = tx().getFieldAmount (sfTakerGets);
+    auto saTakerPays = tx()[sfTakerPays];
```
What's `tx()`? It would be better if it used `ctx_.tx`, which is idiomatic.

ximinez replied:
I think `tx()` is the victim of a weird merge / rebase. It was defined as an accessor to `ctx_.tx`, and seemed to be the standard across the transactors. I would have sworn that it there when I started, but now the diff is showing that I added it, so I can only assume that it got removed somewhere along the way, and one of my rebases silently put it back.

I'll take it out.

ximinez replied:
Ok, this is why I'm confused: https://github.com/ximinez/rippled/commit/c15394c42ad19cb6dee761cf63bb04738b7b5316#diff-afcc1209f06865ece476ddeaee53c363R82

ximinez replied:
Done.

vinniefalco commented on src/ripple/app/tx/impl/CreateOffer.cpp:
```diff
@@ -592,83 +664,65 @@ CreateOffer::applyGuts (ApplyView& view, ApplyView& view_cancel)
+    assert(accountFunds(view, account_, saTakerGets,
+        fhZERO_IF_FROZEN, viewJ) > zero);
+    assert(!cancelSequence ||
+        (uAccountSequenceNext - 1 > *cancelSequence));
```
Is there really a point to these asserts?

nbougalis replied:
They ought to be `throw std::runtime_error` or somesuch (`Throw`?), if we never expect them to trigger. That way, if they do the transaction doesn't process and gives a sensible error instead through the exception handler that wraps TX processing.

vinniefalco replied:
The asserts just duplicate the same logic that is in `preclaim`

ximinez replied:
> The asserts just duplicate the same logic that is in `preclaim`

That's exactly why those asserts are there. I left them to reassure myself that I implemented and called `preclaim` correctly, and once that was done, it seemed that there was no harm in leaving them there, since they won't be included in release code. For that reason, I would be against changing them to `throw`s.

vinniefalco replied:
How do you know that the `assert`s are implemented correctly?

ximinez replied:
As much as I can know that anything is correct, they are copied from the `if`s that used to be there, and logically negated. Most of those should be indicated in the context of the diffs.

vinniefalco replied:
If you know that the `asserts` are correct then couldn't that same rationale be applied to "know" that the implementation of `preclaim` is correct?

ximinez replied:
Like I said above, they are intentionally duplicated as part of the development process. I left them because I didn't think there was a significant cost to some extra asserts. Obviously, there is a cost or we wouldn't still be discussing it, so I'll go through and remove them.

scottschurr replied:
Hmmm.  Personally, I liked the asserts.  They are a form of programming by contract.  Each assert states a precondition that was supposed to be checked previously.  The asserts struck me as quite useful since error checks have been sprinkled hither and yon.

I suppose another approach would be to systematically go through the code and write unit tests that hit the (earlier) returned error that prevents hitting the assert.  Then, if all tests pass, it would hypothetically not be possible to hit one of the asserts.  This approach would be less declarative than the asserts, but would possibly provide more security, since we don't run debug code in production.

Just my feeling.  Take it for what you will.

vinniefalco replied:
"I never saw an assert I didn't like"

vinniefalco replied:
@scottschurr The precondition is "preclaim was called." It should not be simply repeating all the logic of preclaim.

scottschurr replied:
@vinniefalco Actually, I think the preconditions are exactly the ones stated by the asserts.  Whether "preclaim was called" is irrelevant, as long as the necessary validations were performed.  At the moment we expect these checks to be performed by preclaim.  At the rate this code churns that expectation may not be valid in three months.

But, again, I'm not willing to go to the mat on this.  It's just my take.

vinniefalco commented on src/ripple/app/tx/impl/apply.cpp:
```diff
@@ -55,10 +55,82 @@ invoke_preflight (PreflightContext const& ctx)
+    compile-time polymorphism of (presumably) static
+    class functions for Transactor and derived classes.
+*/
+template<class Tx>
```
`template<class T>`

ximinez replied:
I used `Tx` to indicate that it the template class was related to `Transactor`, but if it matters, I'll change it to `T`.

vinniefalco commented on src/ripple/app/tx/impl/apply.cpp:
```diff
@@ -55,10 +55,82 @@ invoke_preflight (PreflightContext const& ctx)
+/*
+    invoke_preclaim<Tx> uses name hiding to accomplish
+    compile-time polymorphism of (presumably) static
+    class functions for Transactor and derived classes.
```
Comment explains what is obvious from code inspection

ximinez replied:
It's not immediately obvious, though, and it was enough to trip up @scottschurr at first (https://github.com/ximinez/rippled/commit/5404e614a895bdaeee5e60a2bef7aca7f647d8a1#commitcomment-13477808). If you're just looking casually, the comment saves the time needed to figure that out.

vinniefalco commented on src/ripple/app/tx/impl/Transactor.h:
```diff
@@ -51,6 +51,66 @@ struct PreflightContext
+public:
+    Application& app;
+    ReadView const& view;
+    TER const preflightResult;
```
There's no point to making members of this class `const`, it just prevents the class from being copy assignable and will eventually lead to annoying warnings.

ximinez replied:
I don't want the class to be copy assignable, and I don't want the members to be mutable.

vinniefalco replied:
Let the caller make the decision. e.g.
```
PreclaimContext const result = ...
```

ximinez replied:
```

ximinez replied:
http://melpon.org/wandbox/permlink/HiOdukqWRT1lOvn1

ximinez replied:
Making the struct `const` when there are non-`const` reference members leaves those members mutable.

http://melpon.org/wandbox/permlink/HiOdukqWRT1lOvn1

vinniefalco commented on src/ripple/app/tx/impl/Transactor.h:
```diff
@@ -51,6 +51,66 @@ struct PreflightContext
+struct PreclaimResult
+{
+public:
+    PreclaimContext const ctx;
```
Why does the result include a context?

ximinez replied:
Mainly as a shortcut to avoid having to copy all the individual members of the context into the result.

vinniefalco commented on src/ripple/app/tx/impl/Transactor.h:
```diff
@@ -88,6 +147,44 @@ class Transactor
+
+    static
+    TER
+    preclaim(PreclaimContext const &ctx)
```
I'm concerned that we went from something that was relatively simple, to a complex proliferation of "results" and "contexts", some of which include each other! Is this the best we can do?

ximinez replied:
The complexity was always there, but it was hidden under the covers of `Transactor::doApply` and its overrides.

vinniefalco replied:
Why does the result have to contain a context?

ximinez replied:
https://github.com/ripple/rippled/pull/1305#discussion_r40846600
> Mainly as a shortcut to avoid having to copy all the individual members of the context into the result.

vinniefalco commented:
This makes explicit some ugliness that was implicit in the old arrangement of the code. But we can live with it for now
:+1:

ximinez commented:
Squashed, rebased, and added one more commit which improves some of the comments on the `pre` functions and removes most (if not all) of the "belt and suspenders" `assert`s.

ximinez commented:
The most notable change is https://github.com/ripple/rippled/commit/eb1ea1c0d2fcd4622ced702709ec315184580143#diff-63e38aef10c237031210901286f42aa2L698
